packages/frontend/src/services/indexer/cache.js

```diff
@@ -0,0 +1,146 @@
+        return new Promise(async (resolve) => {
+            const store = await this.getObjectStore();
+
+            const item = {
+                account: {
+                    id: accountId,
+                    kind,
+                },
+                data
+            };
+
+            const request = store.add(item, IDBCursor.primaryKey);
+            request.onsuccess = resolve;
+        });
+    }
+
+    async _updateRecord(accountId, kind, data) {
+        return new Promise(async (resolve) => {
+            const store = await this.getObjectStore();
+
+            store.openCursor().onsuccess = (event) => {
+                const cursor = event.target.result;
+
+                if (cursor) {
+                    const { account } = cursor.value;
+                    const isFound = account.id === accountId
+                        && account.kind === kind;
+
+                    if (isFound) {
+                        const updatedData = cursor.value;
+                        updatedData.data = data;
+                        const request = cursor.update(updatedData);
+
+                        request.onsuccess = resolve;
+                    } else {
+                        cursor.continue();
+                    }
+                }
+            };
+        });
+    }
+
+    _shouldUpdate(lastTimestamp = 0, timeout) {
+        const time = new Date().getTime();
+
+        return time - lastTimestamp >= timeout;
+    }
+
+    /**
+     * The main idea is save the contract-helper from searching through the entire history of the blockchain.
+     * Each next request, we send the last timestamp, while accumulating data on the client.
+     */
+    async accumulate({ accountId, kind, updater, timeout }) {
+        const record = await this._getRecord(accountId, kind);
+        try {
+            const lastTimestamp = record?.data?.timestamp;
+
+            if (this._shouldUpdate(lastTimestamp, timeout)) {
+                let { version, lastBlockTimestamp, list = [] } = await updater(lastTimestamp);
+
+                const prev = record?.data?.list || [];
+
+                const onlyUniqValues = new Set(list.concat(prev));
+                const updated = {
+                    timestamp: lastBlockTimestamp,
+                    list: Array.from(onlyUniqValues),
+                    version,
+                };
+
+                if (Boolean(record)) {
+                    // If the version is updated on the helper
+                    // we should rescan from the beginning of the blockchain
+                    const isVersionChanged = version !== record.version;
+                    if (isVersionChanged) {
+                        record.timestamp = 0;
+                    }
+
+                    await this._updateRecord(accountId, kind, updated);
+                } else {
+                    await this._addRecord(accountId, kind, updated);
+                }
+
+                return updated.list;
+            }
```

Comment:
We should return the latest record if we did not update

Reply:
Yes, last record returns on line 139 in this case